text/0648-aspects-priority-ordering.md

```diff
@@ -0,0 +1,525 @@
+}
+
+const app = new App();
+const stack = new MyStack(app, 'MyStack');
+
+// Apply the aspect to enable versioning on all S3 Buckets
+Aspects.of(stack).add(new EnableBucketVersioning());
+```
+
+### The Problem with Aspects
+
+The current algorithm for invoking aspects (see invokeAspects in [synthesis.ts](https://github.com/aws/aws-cdk/blob/8b495f9ec157c0b00674715f62b1bbcabf2096ac/packages/aws-cdk-lib/core/lib/private/synthesis.ts#L217))
+does not handle all use cases — specifically, when an Aspect adds a new node to the Construct tree and when Aspects are applied out of order.
+
+#### Aspects that Create New Nodes
+
+For Aspects that create new nodes, inherited Aspects are not always applied to the newly created node or resource as expected. This occurs because the
+algorithm only makes a single pass through the Construct tree, and therefore does not visit the newly added node.
+
+A customer reported this in GitHub issue [#21341](https://github.com/aws/aws-cdk/issues/21341). They have one Aspect that adds Tags to all resources
+in a Stack and another Aspect applied on a construct that creates an S3 Bucket on the parent Stack of that construct. The expected behavior is for
+the new S3 Bucket to inherit the Tags from the parent Stack. However, while the bucket is created, it does not inherit any Tags. Below is the
+customer's reproducible example:
+
+```ts
+import { Aspects, Tags, IAspect, Stack, StackProps } from 'aws-cdk-lib';
+import { Bucket } from 'aws-cdk-lib/aws-s3';
+import { Construct, IConstruct } from 'constructs';
+
+export class AspectsIssueStack extends Stack {
+  constructor(scope: Construct, id: string, props?: StackProps) {
+    super(scope, id, props);
+    
+    // Tags is an Aspect:
+    Tags.of(this).add('test-tag', 'test-value');
+    new MyConstruct(this, 'myConstruct');
+  }
+}
+
+class MyConstruct extends Construct {
+
+  constructor(scope: IConstruct, id: string) {
+    super(scope, id);
+    
+    const stack = Stack.of(scope);
+    const s3Bucket = new Bucket(stack, 'bucket-with-tags');
+    
+    Aspects.of(this).add(new MyAspect());
+  }
+}
+
+class MyAspect implements IAspect {
+  public visit(node: IConstruct): void {
+    // The bucket is added to the parent of the construct where the aspect it was initialized
+    const stack = Stack.of(node);
+    const s3Bucket = new Bucket(stack, 'bucket-without-tags-that-should-have');
+  }
+}
+```
+
+The current algorithm starts at the root and invokes all Aspects on the root level. It then recurses through the tree, visiting each of the root's
+child nodes and invoking each child node's inherited and local Aspects.
+
+See the sequence of Aspect invocation below:
+
+1. Start at root (Stack)
+2. Visit Child (Abstraction)
+3. MyAspect creates NewBucket as another Child at the root level
+4. But traversal has already finished with this level, so New Bucket is never visited again to be Tagged
+
+![Illustration of current Aspect invocation order](../images/AspectsDiagram1.png)
```

Comment:
Thanks for your comments here @DaniloTommasinaTR ! Responding to your points below:

> Shouldn't the "red" aspect also be applied to the newly created S3 bucket (since the "red" aspect is inherited from the root node)?

In this situation, no - since the red aspect is not applied on the root (Stack), but rather to the Abstraction which is a child of the root. The Red aspect creates the New Bucket as a child of the root, not of the Abstraction. If it were applied to the root, then yes, the new Bucket would inherit the red aspect.

> Another issue is related to execution priority of inherited aspects when merged with aspects set on the child nodes.
E.g. use priority 50 for the red aspect. Add a blue aspect with priority 10 on the "Abstraction" node.
What would be the correct execution order ?

There is no `red-a` here - the red aspect is only applied on the Abstraction.

Assuming green gets the default priority of 600, the correct execution order would be:

```
blue-b, blue-c
red-b, red-c
green-a, green-b, green-c, green-d
```

> what if the red-aspect adds the LogBucket and adds also a yellow-aspect with priority 0 to it?

We do not allow nested aspects. The [current algorithm](https://github.com/aws/aws-cdk/blob/8b495f9ec157c0b00674715f62b1bbcabf2096ac/packages/aws-cdk-lib/core/lib/private/synthesis.ts#L240) checks if an aspect was added to a node while invoking another aspect. If it does, then this aspect will not be invoked an we emit a warning - we will continue this behavior in our aspects redesign.

Reply:
> > Shouldn't the "red" aspect also be applied to the newly created S3 bucket (since the "red" aspect is inherited from the root node)?
> 
> In this situation, no - since the red aspect is not applied on the root (Stack), but rather to the Abstraction which is a child of the root. The Red aspect creates the New Bucket as a child of the root, not of the Abstraction. If it were applied to the root, then yes, the new Bucket would inherit the red aspect.
> 

ooh, sorry I picked the wrong line in the md file for this comment, my comments refer to [images/ProposedAspectInvocationOrder.png](https://github.com/aws/aws-cdk-rfcs/blob/sumughan/648/aspects-priority-ordering/images/ProposedAspectInvocationOrder.png)

>> what if the red-aspect adds the LogBucket and adds also a yellow-aspect with priority 0 to it?

> We do not allow nested aspects. The [current algorithm](https://github.com/aws/aws-cdk/blob/8b495f9ec157c0b00674715f62b1bbcabf2096ac/packages/aws-cdk-lib/core/lib/private/synthesis.ts#L240) checks if an aspect was added to a node while invoking another aspect. If it does, then this aspect will not be invoked an we emit a warning - we will continue this behavior in our aspects redesign.

Right, I forgot about that check. Is there a particular reason why this should not be allowed? I understand that it requires the code to run multiple passes of the tree to process dynamically added aspects, but otherwise I don't see the reason, why it should not be allowed.

Reply:
> > ooh, sorry I picked the wrong line in the md file for this comment, my comments refer to [images/ProposedAspectInvocationOrder.png](https://github.com/aws/aws-cdk-rfcs/blob/sumughan/648/aspects-priority-ordering/images/ProposedAspectInvocationOrder.png)
> 
> Ah I see - well in this case, we still wouldn't want the newly created Logging Bucket to inherit the red Aspect. Maybe I can make this more clear in the RFC, but if an Aspect creates a new node, that new node will not inherit the Aspect that created it.
> 
> In this example, if the new Bucket inherited the same aspect which created it, it would recurse infinitely creating new buckets. This behavior is designed to prevent infinite recursion.

In this case there should not be an infinite recursion since the LogBucket is a singleton and as such won't be created multiple times. However this use case is what I reported to not be working in my original bug report.
Let's say you put an Aspect on the root node that enforces blocking public access to S3 buckets. I would expect this aspect to be applied also to the LogBucket.
Other valid use cases could be related to aspects that perform tagging, if I add a tagging aspect on the root node, I would expect it to be run on all nodes including those that have been dynamically added.

This issue manifests itself e.g. within cdk pipelines where aspects are not applied as expected: https://github.com/aws/aws-cdk/blob/6bb142e805fdd754755cc54c31c0e6e7970be7f9/packages/aws-cdk-lib/pipelines/lib/main/pipeline-base.ts#L82

This line adds an Aspect that builds the cdk pipeline (and adds a variety of constructs to the tree) just before synthesis. Aspects placed on the root node won't be applied to the cdk pipeline constructs.
The workaround is to explicitly call [buildPipeline()](https://github.com/aws/aws-cdk/blob/6bb142e805fdd754755cc54c31c0e6e7970be7f9/packages/aws-cdk-lib/pipelines/lib/main/pipeline-base.ts#L129) before processing the Aspects, this will expand the tree before the processing of aspects is triggered and workaround the issue, however this is not the behavior I would expect. Both the "automatic" building of the pipeline and the explicit call to build it, should produce the same result.

> 
> > Right, I forgot about that check. Is there a particular reason why this should not be allowed? I understand that it requires the code to run multiple passes of the tree to process dynamically added aspects, but otherwise I don't see the reason, why it should not be allowed.
> 
> An Aspect creating another Aspect is a weird scenario and I can't imagine a use case for it - we do not allow it as it is probably a mistake on the user's behalf.

I don't think it is a weird one, if we look at the cdk pipelines use-case I mentioned above. If we are dynamically adding complex constructs via Aspects, these complex constructs might come with their own Aspects.
Use-cases that I could think of where this could apply: Company policy requirements for audit logging, you want to enforce logging of a set of actions to be stored to a central S3 bucket. Along with the LogBucket, you would have possibly Lambdas, things that adds entries into the cloud-init scripts for EC2 instances to install/start agents, ...
One would expect the aspects placed on the root node to also be applied to all dynamically added constructs.

For context: This is a very valid use-case like for the company I am employed at.
We are developing a central library to be used by all product delivery teams, the central library is developed in a inner-source setup, we have contributors from various departments, those that create re-usable constructs for complex patterns (e.g. web app with cloud-front, S3, API gateway, ....), those that deal with auditing requirements,  those that create security related aspects (e.g. enforce blocking public access on S3, apply a permission boundary to all IAM roles, ...) and those that deal with financial optimizations (e.g. prefer graviton based instances, ARM64 lambdas, ....).
In an ideal world the various teams could develop and contribute in parallel their own portions of work as independent units, the end-result is a dynamic merge of all the aspects defined by the various contributors.
We should not require team 'security' to keep merging their extensions into team 'audit', things should happen automatically through correct propagation of aspects.

Reply:
We prefer the "ordering-by-priority" solution because it is simple to understand and reason about. "Bigger number means later" is a rule that's easy to explain.

The alternative is running all aspects on the tree in a loop, requiring them to `return true` if they made any changes and `return false` if they didn't, and stopping only when they all return `false`. This puts more burden on aspect authors to "get it right", and to avoid troublesome side effects when they are called multiple times on the same construct (which can and must happen under this model).

We are erring on the side of simplicity here with a priority system.

Can you explain your biggest concern?

* Is it that establishing a global order is impossible?
* Or is it that app builders will have to worry about order?

Because if it's the latter, there are ways around that without scrapping the entire idea of priorities. For example, aspect authors in principle control the priority class (with an option for app builders to override for emergency situations).

Reply:
Hi @DaniloTommasinaTR,

That's good to hear! Let's talk about some details then.

> If the inheritance rule of the aspects is not respected for dynamically added constructs, then bug https://github.com/aws/aws-cdk/issues/21341 will not be solved.

In the current plan: if a construct is added by an aspect with priority `X`, then it will be visited by aspects with priority `>X` (but not necessarily by other aspects with priority `X`).

Is that respected enough in your opinion?

> There is already code in place to ensure that the same aspect is invoked exactly once on the same node

I know. I'm speaking about a hypothetical future scenario where we *wouldn't* do priority-based aspect ordering, but we would do a stabilization loop instead. A stabilization loop is the only alternative to priority-based ordering, and it would have the property that aspects would need to be invoked however often is necessary for them to all stop making modifications.

I thought you were advocating against priorities, which is why I brought up the alternative and its consequences.

> There is no requirement for the Aspect to return true or false, the loop on the tree is repeated as long as at least one Aspect is visited in a full pass of the tree.

Right. We considered this as well, and perhaps this doesn't come through in the doc as well as it should have. Your solution works perfectly well for newly-added constructs: we could detect them and invoke aspects only for constructs we haven't visited yet, and when aspects stop adding constructs we are done.

This does not the (possible) case however of mutations: when an aspects modifies a construct in-place, by calling a method or changing a property, another construct might want to respond to that as well (for example: one aspect sets a property on a bucket, and another aspect tags all buckets with that property set). Since in general there is no way of knowing if a construct was mutated, there is no way for us to detect this.

We would like to settle on a solution that covers both construct addition as well as construct mutation, which leaves us with either a full stabilization loop, or priorities.

We then prefer priorities because of its simplicity.

Reply:
@rix0rrr Thanks for the explanations, I looked again at the pseudo code in the proposed solution and I now understand. I was misinterpreting the collection of the Aspects and mapping to node happening once at the beginning and thought that this would not cover the use-case of inherited aspects. However now I see that it does at least under certain circumstances. The initial mapping just determines to nodes to which aspects have been explicitly added, but then you recurse into each child node, which covers the inheritance use-case.  However the inheritance is valid only for the aspect itself being processed or any aspects with an higher priority.
So, issue # 213411 could be fixed, but only under some quite obscure conditions and you might end up with only part of the inherited Aspects to be processed, which is an "obscure" behavior in my opinion.

The image [images/ProposedAspectInvocationOrder.png](https://github.com/aws/aws-cdk-rfcs/blob/sumughan/648/aspects-priority-ordering/images/ProposedAspectInvocationOrder.png) however does not reflect this behavior and it got me on the wrong track. The aspects being processed is inherited into the dynamically added child (with the risk of endless recursion if the Aspect does not explicit captures the case).

In my opinion there is no way around a stabilization loop, execution order-by-priority is defined per node, if you do not allow aspects adding other aspects, that's feasible. If you would allow aspects adding aspects you might keep execution order if you allow to add aspects only at higher priority values than the currently executed ones. Otherwise you would need to accept some trade-offs on execution order.

Reply:
> The initial mapping just determines to nodes to which aspects have been explicitly added, but then you recurse into each child node, which covers the inheritance use-case. However the inheritance is valid only for the aspect itself being processed or any aspects with an higher priority.

@DaniloTommasinaTR what do you mean by this last sentence? Do you have an example of where this would not work?

I'll update the image, thanks for pointing that out and apologies for the confusion there

Reply:
> One would expect A, B and C to all be processed on the new child following the inheritance rules... at least, I would expect it ;)

I think this is where we have the crux:

The contract we were thinking of is that if aspect `A` adds an aspect `B`, it would only run if `prio(B) > prio(A)`. 

And crucially: **the same would hold whether it would be a new aspect on an existing node, as well as a new aspect on a new node**.

The reason this would be true is so that we don't have a distinction between node additions and node mutations: it's all the same system, which is nice and consistent.

But after thinking about it a little: **sometimes constructs come with their own aspects**, like `Role` and `CdkPipeline`... and I agree with you that it would be unexpected if aspect `A` does `new Role`, but that `new Role` doesn't behave the same as if you did it in another location.

So I think I'm coming around to the idea of **added nodes running all aspects in priority order**, and that **priority ordering is only for the relative ordering of mutating aspects**.

Reply:
@DaniloTommasinaTR thank you for your feedback - we've updated the proposed solution RFC to incorporate the edge cases you pointed out.